Add CEP for Repodata Wheel Support (#145)
conda/ceps: cep-XXXX.md

- The conda-forge community successfully scaled recipe creation

### Wheel Support Proposal (2017)

Issue [#5202](https://github.com/conda/conda/issue/5202) proposed direct wheel installation support, recognizing:

- Wheels provide binary distributions similar to conda packages
- Growing wheel availability on PyPI reduced build complexity
- Could bridge the gap between conda's reliability and PyPI's breadth

### Broader Context (2018)

The Python Discourse thread on [packaging scope boundaries](https://discuss.python.org/t/drawing-a-line-to-the-scope-of-python-packaging/883) reflects the larger ecosystem's struggle with:

- Multiple competing package management tools (pip, conda, poetry, pipenv)
- Unclear responsibilities and interoperability expectations
- Need for clearer standards and communication between tools

### conda-pypi development starts (2022)

Jaime Rodríguez-Guerra starts development on the conda-pypi plugin aimed to improve conda and PyPI interoperability.

### Rip development starts (2023)

Prefix.dev starts a barebones pip implemented in Rust to resolve and install PyPI dependencies with Pixi.

### Pixi Integrates with uv (Jan 2024)

> Pixi changes course to use uv directly instead of rip, which unlocks features like editable installations, and git and path dependencies.

Reviewer: these are all now available for conda-only workflows through pixi-build.

Author: Hey @pavelzw, thanks so much for the feedback! Do you think we are missing a milestone in our brief history section? This pixi build feature is more about building path/git for conda packages than installing wheels isn't it?

### conda-pupa creates on-the-fly conversion plugin (July 2024)

> Daniel Holth creates a conda plugin that supports on-the-fly conversion of conda packages to wheels in a local channel.

Reviewer 2: This seems to be leaving out some history

- Users must understand two package managers, their interaction points, and which of their dependencies are available from which  ecosystem
- Where there is support for mixing environments, it requires multiple solves which reduces overall package installation performance

By adding native support for pure Python wheels to repodata, conda clients can:

- Resolve dependencies across conda and PyPI packages in a single solve
- Provide users with transparent access to the broader Python ecosystem
- Maintain environment consistency and reproducibility
- Eliminate the cognitive burden of managing two package managers
- Fill gaps in conda package availability without requiring new conda builds
- Reduce the maintenance burden by fully or partially eliminating the need to create and maintain conda recipes for pure-Python packages

## Specification

### Add a new top-level key `packages.whl` to list wheels in a channel

According to the current draft schema for [repodata.json][repodata-schema], repodata consists of five top
level keys:

- repodata_version
- info
- packages
- packages.conda
- removed
- signatures

> This CEP proposes the addition of a new `packages.whl` section to account for the wheel format. This key points to a mapping that MUST contain [repodata record][repodata-record-schema] objects. The key of this mapping MUST be the wheel file name, and the contents of the package record MUST contain metadata related to that wheel.

Reviewer 3: ```suggestion
This CEP proposes the addition of a new `packages.whl` section to account for the wheel format. This key points to a mapping that MUST contain [repodata record][repodata-record-schema] objects. The key of this mapping MUST follow the specification described below in "Key naming requirements".
```
Just making sure this part of the specification agrees with the rest.

### Record structure and naming convention

To support sparse repodata processing and maintain compatibility with conda's existing infrastructure, wheel records MUST use conda-style naming conventions rather than wheel filenames as keys. This allows solvers to efficiently filter packages by name without parsing the full package record.

### Key naming requirements

> The key for each entry in `packages.whl` must follow this format:

Reviewer 3: ```suggestion
The key for each entry in `packages.whl` MUST follow this format:
```
Style consistency.

```
{name}-{version}
```

Examples:

- `httpx-0.28.1`
- `typing_extensions-4.15.0`
- `lazy-loader-0.4`

### Why wheel names may differ from conda names

Several factors can cause wheel names to differ from conda-style names:

1. Name normalization: PyPI normalizes dashes to underscores (PEP 503)
  - Example: `lazy-loader` (conda-forge) vs `lazy_loader` (PyPI wheel)
2. Python-specific clarification: PyPI packages are implicitly Python libraries
  - Example: `authzed-py` (conda-forge) vs `authzed` (PyPI)
3. Variant differences: Conda may offer multiple variants with different dependencies
  - Example: `seaborn-base` (conda-forge) vs `seaborn` (PyPI)
4. Cross-channel naming: Different conda channels may use different names

>   - Example: `pyperformance` (conda-forge) vs `performance` (main)

Reviewer 3: ```suggestion
1. **Name normalization:** PyPI normalizes dashes to underscores (PEP 503)
  - Example: `lazy-loader` (conda-forge) vs `lazy_loader` (PyPI wheel)
2. **Python-specific clarification:** PyPI packages are implicitly Python libraries
  - Example: `authzed-py` (conda-forge) vs `authzed` (PyPI)
3. **Variant differences:** Conda may offer multiple variants with different dependencies
  - Example: `seaborn-base` (conda-forge) vs `seaborn` (PyPI)
4. **Cross-channel naming:** Different conda channels may use different names
  - Example: `pyperformance` (conda-forge) vs `performance` (main)
```
Style edits to stay consistent with the rest of the document.

### Naming standard

When there are naming differences between channels, wheel records MUST use the conda-forge package name as the standard. This choice is made because:

- conda-forge is the largest community channel with the most packages
- conda-forge naming conventions are well-established and documented
- Using a single standard prevents ambiguity and tooling complexity
- Channel operators adding wheels to repodata MUST determine the appropriate conda-style name by:
  - Checking if the package already exists in conda-forge and using that exact name
  - If not in conda-forge, applying conda naming conventions per [CEP 26 - Identifying Packages and Channels in the conda Ecosystem][cep-26]
  - Documenting any naming mappings for their channel

### Wheel download URLs

This CEP introduces a new optional `artifact_url` field in package records to specify download locations for individual packages.

> > Note for this draft: The `artifact_url` field could also be added as a separate CEP to allow it for other record types.

Reviewer 3: I think that would actually be a good idea to avoid asymmetries in package record specifications. Either that or we explicitly mention this new field is for all package record types.

- If `artifact_url` is not present (`null`), the download location is constructed from `base_url` and `fn` (existing behavior)

This approach allows packages to be served from:

- A shared relative or absolute `base_url` with all wheels in the same directory, by populating the `base_url` field and leaving the `artifact_url` field empty.
- A manual PyPI repository with wheels in directories by the package name by populating the absolute URL in the `artifact_url` field, or the `base_url` and a relative path in the `artifact_url` field.
- External PyPI mirrors or CDNs using absolute URLs by populating the `artifact_url` field, for example to https://files.pythonhosted.org/packages/.../package-1.0.0-py3-none-any.whl
- Mixed sources within the same repodata file

### Wheel-Specific Record Values

When populating repodata records for pure Python wheels:

- `build: MUST be py3_0
- `build_number`: MUST be 0 for the initial addition of a wheel version. MAY be incremented for subsequent rebuilds of the same - wheel version (e.g., to correct dependencies or metadata)
- `fn`: MUST be the wheel filename (e.g., package-1.0.0-py3-none-any.whl)

> - `subdir`: MUST be noarch

Reviewer 3: ```suggestion
- `subdir`: MUST be "noarch"
```
Consistency

2. Pure Python only: This CEP explicitly does not address wheels with binary extensions, which require platform-specific compatibility guarantees beyond the current scope. Conda’s strength is binary compatibility, so using conda packages may be the optimal solution.
3. Environment markers: Only Python version markers are converted to dependencies. Other environment markers (OS, platform, etc.) are ignored based on the pure Python assumption.
4. Conditionals and Extras: Conditional dependencies and extras are specified by a separate CEP.
5. Repodata size: Supporting a significant portion of pure Python packages from PyPI (potentially hundreds of thousands of packages with multiple versions each) will substantially increase repodata size. Channels adopting wheel support at scale will need to implement sharded repodata ([CEP 16][cep-16]) to maintain acceptable performance. Alternatively, channels may choose to curate a subset of popular or requested packages rather than mirroring all of PyPI.

## Implementation Notes

### For conda clients

> Clients implementing this CEP should:

Reviewer 3: ```suggestion
Clients implementing this CEP SHOULD:
```
Consistency.

This example also demonstrates conditional dependencies. The original `METADATA` file from the wheel has the following dependency information:

```
Requires-Python: >=3.8
Requires-Dist: typing-extensions>=4.0.0; python_version < '3.9'
```

Once conditional dependencies are supported in a separate CEP, the package record should only declare `typing_extensions` as a dependency when the Python version is < 3.9. In this example, the dependency is declared unconditionally for all Python versions. The Python version constraint of >=3.8 is directly mapped.

## Rejected ideas

### Only install Python and pip inside conda environments

In this scenario, users only install Python and pip inside a clean conda environment. Here, we simply use conda as an environment manager and let pip manage the project dependencies.

This is what that typically looks like:

```
conda create -n pip-environment python pip
conda activate pip-environment
pip install <package>
```
Despite its safety for environment management, relying solely on conda for this purpose prevents leveraging the package distribution capabilities of the conda ecosystem.

### Editable installs with conda for dependencies only

> Conda provides all the dependencies of a given package. Then that package is installed on top in editable mode, without addressing dependencies to make sure we don’t accidentally overwrite conda files:

Reviewer 3: ```suggestion
Conda provides all the dependencies of a given package. Then that package is installed on top in editable mode, without addressing dependencies to make sure conda files aren't accidentally overwritten:
```
Writing style nitpick.

```
git clone https://github.com/owner/package.git
conda create -n editable-install package --deps-only
conda activate editable-install
pip install -e . --no-deps
```

### Add more conda packages

Create and maintain new conda packages for each PyPI dependency needed. Tools like [Grayskull] exist to make this easier to convert. However, this is a significant workload for the community, with over half of all conda-forge packages being pure Python. Even with more dedicated resources, creating recipes for over 400 thousand pure Python packages is not achievable.

### Add interoperability to tools through pip dependency scanning

The original version of the conda-pypi plugin called `pip` with the `--dry-run` option to analyze the solution to install a package. With a list of all the dependencies needed, the plugin installed everything available in conda channels first and what was left over would be installed with `pip install --no-deps`. Disadvantages of this approach include:

> - Users still have to know which packages they want from PyPI and then have to `conda pip install` them.

Reviewer 3: ```suggestion
- Users still have to know which packages they want from PyPI and then have to run `conda pip install` to install them.
```
To stay consistent with how it referenced to below.

- Although Pixi and uv are both very fast, it is still slower than performing a single solve of the environment.

### Direct PyPI communication without repodata

Another alternative would be for conda clients to query PyPI's API directly during solving, fetching wheel metadata on-demand rather than including it in repodata. This idea was rejected due to:

- Conda's SAT-based solvers (libsolv, resolvo) require complete package metadata upfront to solve in a single pass, while PyPI's API is designed for lazy evaluation with incremental fetching and backtracking (as pip and uv do). Adapting conda's solvers would require fundamental architectural changes.

- On-demand fetching would also require hundreds of sequential HTTP requests during solving, making solves slow and network-dependent while breaking offline capability and reproducibility guarantees.

### Automatic wheel to conda package conversion and hosting

Another alternative would be establishing a build farm to automatically convert wheels to conda packages and host them on a channel, with conversion triggered by popularity, community requests, or on-demand.

#### Advantages of automatic conversion

> - Unified format: All packages are conda packages, simplifying client implementation

Reviewer 3: ```suggestion
- All packages are conda packages, simplifying client implementation
```
Maybe remove this part to stay consistent with the rest of the list items?

  Travis Hathaway &lt;travis.j.hathaway@gmail.com&gt;
</td></tr>
<tr><td> Created </td><td> Dec 23, 2025</td></tr>
<tr><td> Updated </td><td> Dec 23, 2025</td></tr>
<tr><td> Discussion </td><td> https://github.com/conda/ceps/pull/ </td></tr>
<tr><td> Implementation </td><td> TBD </td></tr>
<tr><td> Requires </td>N/A</tr>
</table>

> The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT",
"RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as
described in [RFC2119][RFC2119] when, and only when, they appear in all capitals, as shown here.

## Abstract

This CEP outlines how native support for pure Python wheel packages will be achieved by adding support for them to conda's package index (repodata). When implemented, conda clients will be able to seamlessly install conda packages and pure Python wheels from enabled channels.
We explicitly limit the scope of this CEP to pure Python wheels to avoid platform-specific binary compatibility issues, but even with this limitation, we dramatically expand the number of available packages for users in the conda ecosystem.

## Motivation

> While conda remains a language-agnostic packaging distribution, installing packages for use with Python represents the majority of conda usage. Users frequently encounter packages only available as wheels on PyPI, forcing them to use hybrid workflows that mix conda and pip installations. This approach creates several problems:

Reviewer 4: ```suggestion
While conda remains a language-agnostic packaging distribution, installing packages for use with Python represents the majority of conda usage. Users frequently encounter packages only available as wheels on PyPI, forcing them to use hybrid workflows that mix conda and PyPI installations. This approach creates several problems:
```

- Fill gaps in conda package availability for simpler pure-Python packages
- Reduce the maintenance burden for straightforward pure-Python packages that don't require metadata modifications

Note that this CEP does not eliminate the need for conda recipes entirely. Many packages will still require proper conda feedstocks due to metadata differences, dependency corrections, or other ecosystem-specific needs. This CEP provides an additional tool for handling simpler pure-Python cases more efficiently.

## Specification

### Add a new top-level key `packages.whl` to list wheels in a channel

According to the current draft schema for [repodata.json][repodata-schema], repodata consists of five top
level keys:

- repodata_version
- info
- packages
- packages.conda
- removed
- signatures

This CEP proposes the addition of a new `packages.whl` section to account for the wheel format. This key points to a mapping that MUST contain [repodata record][repodata-record-schema] objects. The key of this mapping MUST follow the specification described below in [Key naming requirements](#key-naming-requirements).

### Record structure and naming convention

> To support sparse repodata processing and maintain compatibility with conda's existing infrastructure, wheel records MUST use conda-style naming conventions rather than wheel filenames as keys. This allows solvers to efficiently filter packages by name without parsing the full package record.

Author: Thanks, I updated this section.

- `{build}` is the build string (e.g., `py3_0`, `py3_1`) from the `build` field
- `{abi_tag}-{platform_tag}` are extracted from the wheel filename (stored in the `fn` field)

Examples:

- `httpx-0.28.1-py3_0-none-any`
- `typing_extensions-4.15.0-py3_0-none-any` (METADATA name `typing-extensions` mapped to conda name `typing_extensions`)
- `requests-2.32.5-py3_1-none-any` (rebuild with build_number 1)

### Why wheel names may differ from conda names

Several factors can cause wheel names to differ from conda-style names:

1. **Wheel filename normalization:** Wheel filenames normalize hyphens to underscores per PEP 427, but the METADATA `Name` field contains the canonical name (which may have hyphens).
    - Example: `lazy-loader` (METADATA name) vs `lazy_loader` (wheel filename)

2. **Python-specific clarification:** PyPI packages are implicitly Python libraries
    - Example: `authzed-py` (conda-forge) vs `authzed` (PyPI)

3. **Variant differences:** Conda may offer multiple variants with different dependencies
    - Example: `seaborn-base` (conda-forge) vs `seaborn` (PyPI)

4. **Cross-channel naming:** Different conda channels may use different names

>     - Example: `pyperformance` (conda-forge) vs `performance` (main)

Author: Thanks, I moved this to a rationale section.

### Naming standard and channel mapping

> The key name for wheel records SHALL be constructed by combining the conda-style name (derived from METADATA and normalized per conda conventions) with the version, build string, and wheel tags (abi_tag-platform_tag) extracted from the filename. The python_tag is omitted as it is redundant with the build string. This ensures the name portion follows conda naming conventions, supports multiple rebuilds, and the tags portion aligns with the wheel filename structure.

Reviewer 4: I don't see how we can extend the name-version-build syntax with extra fields without hurting backwards compatibility. All the wheel tags should be condensed in the build string, or dropped from the key.

Author: Hi @jaimergp, are you saying that we could limit client changes by making what was the "build" string a much longer set of wheel tags?

When there are naming differences between PyPI wheels and conda packages, channel operators MUST determine the appropriate conda-style name by applying conda naming conventions per [CEP 26 - Identifying Packages and Channels in the conda Ecosystem][cep-26].

To help users understand which naming conventions are being used, channels MAY reference a parent channel that defines the naming conventions and mappings for wheel-to-conda name translation. Parent/child channel relationships will be defined in a separate CEP, which will specify how channels can declare a `parent_channel` field in the `info` section of repodata to inherit naming conventions and mappings from another channel.
For example, a wheel channel could reference conda-forge as its parent channel, inheriting conda-forge's established naming conventions and package name mappings. This allows channels to avoid duplicating name mapping definitions and ensures consistency across related channels.

> Channel operators SHOULD document any naming conventions and mappings specific to their channel, including whether they reference a parent channel for naming conventions.

Author: Hi @jaimergp, we need to map the names of all the packages and the names of all the depends, otherwise the solver won't be able to create a solution. The PyPI name is saved as part of the fn field for the filename.

### Wheel download URLs

> This CEP introduces a new optional `artifact_url` field in package records to specify download locations for individual packages.

Reviewer 3: I don't have any problems with that.

> This CEP introduces a new optional `artifact_url` field in package records to specify download locations for individual packages.

Author: Thanks! I renamed this to `url`.

The build portion (`py3_0`) comes from the `build` field. The tags portion (`none-any`) is extracted from the wheel filename (`annotated_types-0.7.0-py3-none-any.whl`), which normalizes the package name to underscores per PEP 427.
2. Dependency name mapping: This package depends on `typing_extensions`, which is listed in the `depends` field. On PyPI, this package is named `typing-extensions` (with a hyphen), but it has been mapped to the name `typing_extensions` (with an underscore) to match the existing conda-forge package name. Such mappings may be inherited from a parent channel (see [Naming standard and channel mapping](#naming-standard-and-channel-mapping)).

This example also demonstrates conditional dependencies. The original `METADATA` file from the wheel has the following dependency information:

```text
Requires-Python: >=3.8
Requires-Dist: typing-extensions>=4.0.0; python_version < '3.9'
```

The package record uses conditional dependency syntax (`; if python < 3.9`) to declare `typing_extensions` only when the Python version is < 3.9, matching the original wheel METADATA. The Python version constraint of >=3.8 is directly mapped.

> ## Rejected ideas

Author: Nice catch, I forgot about that one. I added it as a rejected approach, thanks!

4. **Cross-channel naming:** Different conda channels may use different names
    - Example: `pyperformance` (conda-forge) vs `performance` (main)

### Naming standard and channel mapping

When there are naming differences between PyPI wheels and conda packages, channel operators MUST determine the appropriate conda-style name by applying conda naming conventions per [CEP 26 - Identifying Packages and Channels in the conda Ecosystem][cep-26].

To help users understand which channel's naming conventions are being used, channels MAY declare an optional `name_mapping_channel` field in the `info` section of repodata. This field indicates which channel's package names are used as the standard for wheel-to-conda name mapping in this repodata.

```json
{
  "info": {
    "subdir": "noarch",
    "base_url": "https://repo.example.com/channel/",
    "name_mapping_channel": "conda-forge"
  }
}
```

When `name_mapping_channel` is present, channel operators SHOULD:

- Use the declared channel's package names as the standard for name mapping
- Document any naming mappings specific to their channel
- Ensure consistency with the declared channel's naming conventions

> Channel operators MUST document any naming conventions and mappings specific to their channel, regardless of whether `name_mapping_channel` is declared.

Reviewer 5: Are we confortable with saying `MUST` here, considering the current state of affairs when it comes to mappings in conda-forge and main? (and to be fair with conda-forge, they are in a better state than main).

Author: Thanks, I updated it from MUST to SHOULD.

If a centralized name mapping standard becomes available (such as the external dependency registry proposed in [PEP 804][pep-804]), channels MAY use that standard for name mapping instead of or in addition to channel-specific conventions. This would provide a unified approach to name mapping across ecosystems and reduce discrepancies.

### Wheel download URLs

This CEP introduces a new optional `artifact_url` field in package records to specify download locations for individual packages.

> Note for this draft: The `artifact_url` field could also be added as a separate CEP to allow it for other record types.

When present, the `artifact_url` field follows these semantics:

- If `base_url` is defined in the repodata info object (per [CEP 15][cep-15]), `artifact_url` contains the path relative to `base_url`
- If `base_url` is not defined, `artifact_url` contains the complete download URL

> - If `artifact_url` is not present (`null`), the download location is constructed from `base_url` and `fn` (existing behavior)

Reviewer 5: Minor: This is a bullet point under "When present, the `artifact_url` field follows these semantics". You could likely do this:
>When present, the `artifact_url` field follows these semantics:
>- If `base_url` is defined in the repodata info object (per [CEP 15][cep-15]), `artifact_url` contains the path relative to `base_url`
>- If `base_url` is not defined, `artifact_url` contains the complete download URL
>
>When not present, the download location is constructed from `base_url` and `fn` (existing behavior)

?

Author: Thanks so much for this feedback, I updated this based on your suggestion.

Solvers MUST treat pure Python wheels as valid package candidates during dependency resolution with these constraints:

- **Exclusivity:** Solvers MUST NOT install both a conda package and wheel for the same package name.
- **Dependency satisfaction:** When a wheel is selected, its `depends` list MUST be satisfied like any conda package.
- **Platform matching:** Since all wheels in `packages.whl` are pure Python (noarch), no platform filtering is needed.

#### User control of precedence

By default, channels with conda packages MUST prefer conda packages when they are available. Users MAY override default precedence through:

- Channel priority configuration (prefer channels with wheels).
- Explicit wheel requests through `<channel name>::<package name>` syntax.
- Explicit configuration in the client itself (e.g. `prefer_conda` or `prefer_wheel`)

### Limitations

This CEP has the following known limitations:

1. **Pure Python only:** This CEP explicitly does not address wheels with binary extensions, which require platform-specific compatibility guarantees beyond the current scope. Conda’s strength is binary compatibility, so using conda packages may be the optimal solution.

> 2. **Environment markers**:** Only Python version markers are converted to dependencies. Other environment markers (OS, platform, etc.) are ignored based on the pure Python assumption.

Author: Hi @JeanChristopheMorinPerso, since conditional dependencies is a separate CEP, what if we said that channel operators MAY patch and that they SHOULD support further conditional dependencies as they become available?

> 2. **Environment markers**:** Only Python version markers are converted to dependencies. Other environment markers (OS, platform, etc.) are ignored based on the pure Python assumption.

Reviewer 6: I think its important that we do describe how to convert those markers when conditional dependencies do exist. Because not all markers are trivial to convert (e.g. `platform_machine`, `python_full_version` vs `python_version`). I think that should be in this CEP but I can also imagine that a following could work.

> 2. **Environment markers**:** Only Python version markers are converted to dependencies. Other environment markers (OS, platform, etc.) are ignored based on the pure Python assumption.

Author: I added a non-normative reference to this for our conda-pypi conversion table.

3. **Conditionals and Extras:** Conditional dependencies and extras are specified by a separate CEP.
4. **Repodata size:** Supporting a significant portion of pure Python packages from PyPI (potentially hundreds of thousands of packages with multiple versions each) will substantially increase repodata size. Channels adopting wheel support at scale will need to implement sharded repodata ([CEP 16][cep-16]) to maintain acceptable performance. Alternatively, channels may choose to curate a subset of popular or requested packages rather than mirroring all of PyPI.

> 5. **PyPI package deletion:** PyPI allows package maintainers to delete packages (as opposed to just yanking them), which can break locked environments that reference those packages. Channels using external PyPI URLs directly are subject to this risk. For production use and reproducible environments, channels SHOULD mirror and store wheel artifacts locally rather than relying solely on external PyPI URLs. This ensures that packages referenced in lock files remain available even if deleted from PyPI.

Reviewer 5: I agree that deletion is sometimes a problem, but I'm wondering if this could encourage unsafe practices. Maybe we should add some wording or warnings regarding security. Or we could also remove the recommendation altogether or replace
>For production use and reproducible environments, channels SHOULD mirror and store wheel artifacts locally rather than relying solely on external PyPI URLs. This ensures that packages referenced in lock files remain available even if deleted from PyPI.

with
>For production use and reproducible environments, channels can mirror and store wheel artifacts locally rather than relying solely on external PyPI URLs.

(replace should with can, and remove the last sentence).

Opinions?

Author: Thanks for the feedback, I like those wording clarifications. I went with "MAY mirror" to match the keywords.

## Implementation Notes

### For conda clients

Clients implementing this CEP SHOULD:

- Parse the new `packages.whl` section alongside existing package sections
- Apply the same filtering and preference logic used for conda packages
- Extract wheel metadata during solving to populate dependency information
- Provide the ability to natively install wheels or on-the-fly convert wheels to conda packages for installation

### For channel operators

Channel operators adding wheel support should:

- Implement validation to ensure only pure Python wheels are included
- Maintain a mapping of PyPI to conda-style names for their channel

> - Consider automation to keep wheel metadata synchronized with PyPI

Reviewer 5: I don't know if this is part of a standard or not, but files on PyPI are immutable. So it's impossible to modify the metadata of an existing wheel. Is it possible that you meant "automation to keep the repodata up-to-date with newer releases on PyPI"?

This approach allows packages to be served from:

- A shared relative or absolute `base_url` with all wheels in the same directory, by populating the `base_url` field and leaving the `artifact_url` field empty.
- A manual PyPI repository with wheels in directories by the package name by populating the absolute URL in the `artifact_url` field, or the `base_url` and a relative path in the `artifact_url` field.
- External PyPI mirrors or CDNs using absolute URLs by populating the `artifact_url` field, for example to <https://files.pythonhosted.org/packages/.../package-1.0.0-py3-none-any.whl>
- Mixed sources within the same repodata file

### Wheel-Specific Record Values

When populating repodata records for pure Python wheels:

- `build`: MUST be py`PY_MAJOR_VERSION`_`build_number` (e.g. py3_0)
- `build_number`: MUST be 0 for the initial addition of a wheel version. MAY be incremented for subsequent rebuilds of the same - wheel version (e.g., to correct dependencies or metadata)
- `fn`: MUST be the wheel filename (e.g., package-1.0.0-py3-none-any.whl)
- `subdir`: MUST be "noarch"
- `noarch`: MUST be "python"

> - `artifact_url`: MUST be present and follow the semantics described above

Author: Thanks, changed it to SHOULD.

- Adjust version constraints for conda ecosystem compatibility
- Correct dependency name mappings that were incorrectly converted
- Add or modify metadata fields as needed

This patching capability is essential for:

- **Quality control:** Test and validate packages before exposing to users
- **Ecosystem needs:** Resolve naming conflicts and dependency incompatibilities
- **Reproducibility:** Optionally mirror and store wheel artifacts locally to prevent external deletions
- **Downstream fixes:** Address metadata issues that would break environments

Channels SHOULD document their patching policies and maintain transparency about which packages have been modified from their upstream PyPI metadata.

### Dependency conversion

Wheel dependencies (from METADATA file's Requires-Dist entries) MUST be converted to conda format following these rules:

- **Package names:** Names per [CEP 26][cep-26] and match existing conda-forge package names where they exist
- **Version specifiers:** Map PEP 440 [version specifiers][version-specifiers] to conda format:
  - `==X.Y.Z` → `X.Y.Z` (exact pin)
  - `>=X.Y.Z` → `>=X.Y.Z`
  - `<=X.Y.Z` → `<=X.Y.Z`
  - `<X.Y.Z` → `<X.Y.Z`
  - `>X.Y.Z` → `>X.Y.Z`

>   - `~=X.Y` → `>=X.Y,<X+1.0` (compatible release)

Reviewer 6: Conda also has this specifier.

Author: Thanks, TIL! I updated the proposal to include the `~=`.

  - `!=X.Y.Z` → Add to `constrains` field
- **Multiple specifiers:** Combine with commas (e.g., >=1.0,<2.0)
- **Python version requirements:** Convert Requires-Python to explicit python dependency

> - **Environment markers:** Ignore markers other than Python version (pure Python assumption)

Author: I updated the CEP to allow for additional environment markers. I also referenced a non-normative conversion table from conda-pypi.

Example conversion:

```text
# Wheel METADATA
Requires-Python: >=3.8
Requires-Dist: requests (>=2.20.0,<3.0.0)
Requires-Dist: click (>=7.0)
Requires-Dist: numpy (>=1.20.0,!=1.24.0)
Requires-Dist: importlib-metadata (>=1.0) ; python_version < '3.8'
```

Resulting conda record:

```json
{
  "depends": [
    "python >=3.8",
    "requests >=2.20.0,<3.0",
    "click >=7.0",
    "numpy >=1.20.0"
  ],
  "constrains": [
    "numpy !=1.24.0"
  ]
}
```

> > **Note:** The `importlib-metadata` dependency is omitted because the `Requires-Python: >=3.8 makes its environment marker always false because the package requires Python >=3.8, making the python_version < '3.8' marker always false.

Reviewer 6: That's lucky in this case, maybe just remove this?

### Handling conditional and extra dependencies

Like in the example above of only requiring `importlib-metadata` for certain Python versions, conditional and extra dependencies MUST be supported to enable full interoperability between the ecosystems. This will be supported through a separate CEP: <https://github.com/conda/ceps/pull/111>.

### Solver behavior and package preference

#### Dependency resolution

Solvers MUST treat pure Python wheels as valid package candidates during dependency resolution with these constraints:

- **Exclusivity:** Solvers MUST NOT install both a conda package and wheel for the same package name.

> - **Dependency satisfaction:** When a wheel is selected, its `depends` list MUST be satisfied like any conda package.

Author: Thanks, I added this clarification!

4. **Repodata size:** Supporting a significant portion of pure Python packages from PyPI (potentially hundreds of thousands of packages with multiple versions each) will substantially increase repodata size. Channels adopting wheel support at scale will need to implement sharded repodata ([CEP 16][cep-16]) to maintain acceptable performance. Alternatively, channels may choose to curate a subset of popular or requested packages rather than mirroring all of PyPI.
5. **PyPI package deletion:** PyPI allows package maintainers to delete packages (as opposed to just yanking them), which can break locked environments that reference those packages. Channels using external PyPI URLs directly are subject to this risk. For production use and reproducible environments, channels MAY mirror and store wheel artifacts locally rather than relying solely on external PyPI URLs.

## Implementation Notes

### For conda clients

Clients implementing this CEP SHOULD:

- Parse the new `packages.whl` section alongside existing package sections
- Apply the same filtering and preference logic used for conda packages
- Extract wheel metadata during solving to populate dependency information

> - Provide the ability to natively install wheels or on-the-fly convert wheels to conda packages for installation

Author: This a great point, we have been more focused on getting the repodata to the right format and less on the native installation because our first phase is with conda-pypi which converts wheels to conda packages on the fly. A future step would be to support them natively. I'll start to work on a section about how this will work.

> - Provide the ability to natively install wheels or on-the-fly convert wheels to conda packages for installation

Author: Hey @baszalmstra and @jaimergp, I broke up the CEP in to 3:

1. Informational overview CEP about an overview, the history, and rejected ideas
2. CEP about repodata for wheels
3. CEP about client solving, mapping, and installation

Please let me know what you think.

### For channel operators

Channel operators adding wheel support SHOULD:

- Implement validation to ensure only pure Python wheels are included
- Maintain a mapping of PyPI to conda-style names for their channel
- Consider automation to keep the repodata up to date with newer releases on PyPI
- Document any naming conventions specific to their channel

> - For production channels, consider mirroring wheel artifacts locally to ensure reproducibility and protect against PyPI deletions

Reviewer 4: I'm also thinking about this. We are adding a ton of metadata complexity to support the case where "maybe we don't want to copy this file over in a certain way". At that point, why not create a hybrid solver that knows how to traverse the PyPI index with certain rules (consider only pure Python wheels with dependencies already available, or only pure Python).

This proposal is a bit in the middle of nowhere: it requires preparsing all the wheel metadata and exposing it in a conda-friendly way, but then all the work is left for the client when it comes to installing the wheel. So we either go full conda and convert everything from the beginning, or stay 100% in PyPI world and integrate the solvers better (the mapping problem is present in all three solutions).

This proposal does one thing correctly though: it brings metadata control (by allowing repodata patching) without forcing anyone to copy any files (but in some cases it might be unavoidable).

## Rejected ideas

### Only install Python and pip inside conda environments

In this scenario, users only install Python and pip inside a clean conda environment. Here, we simply use conda as an environment manager and let pip manage the project dependencies.

This is what that typically looks like:

```bash
conda create -n pip-environment python pip
conda activate pip-environment
pip install <package>
```

Despite its safety for environment management, relying solely on conda for this purpose prevents leveraging the package distribution capabilities of the conda ecosystem.

### Editable installs with conda for dependencies only

Conda provides all the dependencies of a given package. Then that package is installed on top in editable mode, without addressing dependencies to make sure conda files aren't accidentally overwritten:

```bash
git clone https://github.com/owner/package.git
conda create -n editable-install package --deps-only
conda activate editable-install
pip install -e . --no-deps
```

### Add more conda packages

> Create and maintain new conda packages for each PyPI dependency needed. Tools like [Grayskull] exist to make this easier to convert. However, this is a significant workload for the community, with over half of all conda-forge packages being pure Python. Even with more dedicated resources, creating recipes for over 400 thousand pure Python packages is not achievable.

Reviewer 4: I could see a `conda-forge/wheels-index` repository where the automation pipelines are maintained folks add their requests to:

1. Add a project to the watchlist so it is included in the index
2. Yank certain wheels
3. Repodata patch
4. Import and archive pure wheel feedstocks
5. etc

I don't know if the resulting repodata would be added to conda-forge proper, or maybe a separate conda-forge-wheels channel (to keep the production channel lighter).

### Automatic wheel to conda package conversion and hosting

Another alternative would be establishing a build farm to automatically convert wheels to conda packages and host them on a channel, with conversion triggered by popularity, community requests, or on-demand.

#### Advantages of automatic conversion

- All packages are conda packages, simplifying client implementation
- Leverages mature conda infrastructure and tooling without client changes
- Enables metadata enrichment, quality control, and validation before publishing

#### Disadvantages and why native wheel support was chosen

Despite these advantages, this approach was rejected because:

- **Infrastructure burden:** Requires significant storage and bandwidth to host and serve converted packages that duplicate PyPI's CDN infrastructure, plus ongoing maintenance of the conversion pipeline and synchronization automation
- **Resource inefficiency:** Wheels are already an excellent format for pure Python packages; conversion adds no technical value and wastes resources

> - **Timeliness:** Creates lag between PyPI publication and availability, with curation decisions needed about which packages to convert

Reviewer 6: This is not different from conversion of the wheels. A manual/automated step still needs to happen to convert the metadata. Converting the wheel data itself takes seconds at worst.

Author: As you point out, wheel conversion could also be automated without curation. I agree that this isn't a major differentiator between the approaches and I removed it. Thanks!

- **Maintenance complexity:** Every wheel format change requires tooling updates; bugs affect all converted packages

> - **Ecosystem misalignment:** Converting away from wheels moves conda further from Python packaging standards

Reviewer 6: Worst case we keep the status quo in my eyes. Direct conversion from a pypi wheel to a conda package already removes the in between step of requiring a recipe.

## Specification

### Add a new top-level key `packages.whl` to list wheels in a channel

According to the current draft schema for [repodata.json][repodata-schema], repodata consists of five top
level keys:

- repodata_version
- info
- packages
- packages.conda
- removed
- signatures

This CEP proposes the addition of a new `packages.whl` section to account for the wheel format. This key points to a mapping that MUST contain [repodata record][repodata-record-schema] objects. The key of this mapping MUST follow the specification described below in [Key naming requirements](#key-naming-requirements).

### Record structure and naming convention

To support sparse repodata processing and maintain compatibility with conda's existing infrastructure, wheel records MUST use conda-style naming conventions rather than wheel filenames as keys. This allows solvers to efficiently filter packages by name without parsing the full package record.

### Key naming requirements

The key for each entry in `packages.whl` MUST follow the format `{name}-{version}`.

Examples:

> - `httpx-0.28.1`

Reviewer 7: I am a little bit afraid that this convention kinda closes the door on other wheel variants (like non-noarch packages) in the future.  We could just align it closer with the actual filename to future-proof things?

> Note for this draft: The `artifact_url` field could also be added as a separate CEP to allow it for other record types.

When present, the `artifact_url` field SHALL follow these semantics:

- If `base_url` is defined in the repodata info object (per [CEP 15][cep-15]), `artifact_url` contains the path relative to `base_url`
- If `base_url` is not defined, `artifact_url` contains the complete download URL

When not present (`null`), the download location is constructed from `base_url` and `fn` (existing behavior).

This approach allows packages to be served from:

- A shared relative or absolute `base_url` with all wheels in the same directory, by populating the `base_url` field and leaving the `artifact_url` field empty.
- A manual PyPI repository with wheels in directories by the package name by populating the absolute URL in the `artifact_url` field, or the `base_url` and a relative path in the `artifact_url` field.
- External PyPI mirrors or CDNs using absolute URLs by populating the `artifact_url` field, for example to <https://files.pythonhosted.org/packages/.../package-1.0.0-py3-none-any.whl>

> - Mixed sources within the same repodata file

Author: This is now a separate CEP at https://github.com/conda/ceps/pull/151.
